components/utils.py

```diff
@@ -520,3 +622,356 @@ def untar(source, destination='/tmp', strip=1):
+        return os.path.join(self.BASE_PROPERTIES_PATH,
+                            service_name,
+                            self.NODE_PROPS_DIR_NAME)
+
+    def _get_rollback_properties_dir(self, service_name):
+        return os.path.join(self.BASE_PROPERTIES_PATH,
+                            service_name,
+                            self.ROLLBACK_NODE_PROPS_DIR_NAME)
+
+
+class BlueprintResourceFactory(object):
+
+    BASE_RESOURCES_PATH = '/opt/cloudify'
+    RESOURCES_DIR_NAME = 'resources'
+    RESOURCES_ROLLBACK_DIR_NAME = 'resources_rollback'
+    RESOURCES_JSON_FILE = '__resources.json'
+
+    UPGRADE_RESOURCE_PREFIX = 'upgrade-'
+
+    def create(self, source, destination, service_name, user_resource=False,
+               source_resource=False, render=True, load_ctx=True):
+        """
+        a Factory used to create a local copy of a resource upon deployment.
+        This copy allows to later reuse the resource for upgrade/rollback
+        purposes.
+        :param source: The resource source url
+        :param destination: Resource destination path
+        :param service_name: used to retrieve node properties to be rendered
+        into a resource template
+        :param user_resource: Resources that should potentially remain
+        identical upon upgrade such as custom security configuration files.
+        These file will be reused provided the ctx node property:
+        use_existing_on_upgrade set to True.
+        :param source_resource: Source resources are source packages that
+        should be downloaded with no rendering and only archived.
+        :param render: Set to false if resource does not require rendering.
+        :param load_ctx: Set to false if properties are not available in the
+        context of the script.
+        :return: The local resource file path and destination.
+        """
+        resource_name = os.path.basename(destination)
+        # The local path is decided according to whether we are in upgrade
+        local_resource_path = self._get_resource_file_path(service_name,
+                                                           resource_name,
+                                                           is_upgrade)
+
+        if not os.path.isfile(local_resource_path):
+            mkdir(os.path.dirname(local_resource_path))
+            if user_resource:
+                self._download_user_resource(source,
+                                             local_resource_path,
+                                             resource_name,
+                                             service_name,
+                                             render=render,
+                                             load_ctx=load_ctx)
+            elif source_resource:
+                self._download_source_resource(source,
+                                               local_resource_path)
+            elif render:
+                self._download_resource_and_render(source,
```

Comment:
since render passes to `download_user_resource`... is there a reason to also have it here?